cfi/lib/src/cfi_ctr.rs

```diff
@@ -0,0 +1,223 @@
+    /// Returns the "default value" for a type.
+    fn default() -> Self {
+        Self::encode(0)
+    }
+}
+
+fn prng() -> &'static Xoshiro128 {
+    use caliptra_drivers::memory_layout::CFI_XO_S0_ORG;
+    if cfg!(feature = "cfi-test") {
+        static mut STATE: Xoshiro128 = Xoshiro128::new_unseeded();
+        unsafe { &STATE }
+    } else {
+        unsafe { Xoshiro128::from_address(CFI_XO_S0_ORG) }
+    }
+}
+
+/// CFI counter
+pub enum CfiCounter {}
+
+impl CfiCounter {
+    /// Reset counter
+    #[inline(never)]
+    pub fn reset(trng: &mut caliptra_drivers::Trng) {
+        prng().seed_from_trng(trng);
+        Self::reset_internal();
+    }
+
+    #[cfg(feature = "cfi-test")]
+    pub fn reset_for_test() {
+        Self::reset_internal()
+    }
+
+    fn reset_internal() {
+        Self::write(CfiInt::default());
+    }
+
+    // Zero the counter
+    pub fn corrupt() {
+        Self::write(CfiInt {
+            val: 0,
+            masked_val: 0,
+        });
+    }
+
+    /// Increment counter
+    #[inline(never)]
+    pub fn increment() -> CfiInt {
+        if cfg!(all(feature = "cfi", feature = "cfi-counter")) {
+            let int = Self::read();
+            if !int.is_valid() {
+                cfi_panic(CfiPanicInfo::CounterCorrupt);
+            }
+
+            let (new, overflow) = int.val.overflowing_add(1);
+            if overflow {
+                cfi_panic(CfiPanicInfo::CounterOverflow);
+            }
+
+            let new_int = CfiInt::encode(new);
+            Self::write(new_int);
+
+            int
+        } else {
+            CfiInt::default()
+        }
+    }
+
+    /// Decrement Counter
+    #[inline(never)]
+    pub fn decrement() -> CfiInt {
+        if cfg!(all(feature = "cfi", feature = "cfi-counter")) {
+            let val = Self::read();
+            if !val.is_valid() {
+                cfi_panic(CfiPanicInfo::CounterCorrupt);
+            }
+
+            let (new, underflow) = val.val.overflowing_sub(1);
+            if underflow {
+                cfi_panic(CfiPanicInfo::CounterUnderflow);
+            }
+
+            let new_val = CfiInt::encode(new);
+            Self::write(new_val);
+
+            Self::read()
+        } else {
+            CfiInt::default()
+        }
+    }
+
+    /// Assert the counters are equal
+    #[inline(never)]
+    pub fn assert_eq(val1: CfiInt, val2: CfiInt) {
+        if cfg!(all(feature = "cfi", feature = "cfi-counter")) {
+            if !val1.is_valid() {
+                cfi_panic(CfiPanicInfo::CounterCorrupt);
+            }
+            if !val2.is_valid() {
+                cfi_panic(CfiPanicInfo::CounterCorrupt);
+            }
+            if val1 != val2 {
+                cfi_panic(CfiPanicInfo::CounterMismatch);
+            }
+        }
+    }
+
+    #[inline(never)]
+    pub fn delay() {
+        #[cfg(all(target_arch = "riscv32", feature = "cfi", feature = "cfi-counter"))]
+        unsafe {
+            let cycles = prng().next() % 256;
```

Comment:
Nitpick: Can you move these first two lines outside the unsafe block? It makes it easier for reviewers to be confident that prng().next() isn't an unsafe method with additional safety requirements.

Reply:
Will do it as part of the asymmetrical increment and decrement calls in the proc macro.